feat: validate nRF Cloud API response

pudkrong opened the pull request:
## PR Type:
Enhancement

___
## PR Description:
This PR introduces validation for the nRF Cloud API responses. It uses the AJV (Another JSON Schema Validator) library and the TypeBox library to define and validate the schema of the responses. In case of validation errors, a ValidationError is thrown. The PR affects multiple files, mainly focusing on implementing the validation in the fetch functions and updating the function calls accordingly.

___
## PR Main Files Walkthrough:
<details> <summary>files:</summary>

`lambda/loggingFetch.ts`: Added a ValidationError class and a validate function to validate the schema of the responses. Also, introduced a new function validatedLoggingFetch which validates the response before returning it.
`lambda/resolveSingleCellGeoLocation.ts`: Replaced loggingFetch with validatedLoggingFetch and updated the function calls to handle the validation errors.
`nrfcloud/createAccountDevice.ts`: Replaced the fetch call with a call to validatedFetch and handled the validation errors.
`nrfcloud/devices.ts`: Replaced the fetch call with a call to validatedFetch and handled the validation errors.
`nrfcloud/getDeviceShadowFromnRFCloud.ts`: Replaced the fetch call with a call to validatedFetch and handled the validation errors.
`nrfcloud/validatedFetch.ts`: Introduced a new function validatedFetch which validates the response before returning it.
</details>

pudkrong committed:
feat: validate nRF Cloud API response

coderbyheart commented on lambda/resolveSingleCellGeoLocation.ts:
```diff
@@ -41,7 +40,7 @@ const deviceFetcher = getDeviceAttributesById({ db, DevicesTableName })
 const { track, metrics } = metricsForComponent('singleCellGeo')
 
-const trackFetch = loggingFetch({ track, log })
+const trackFetch = validatedLoggingFetch({ track, log })
```
```suggestion
const trackFetch = validatedFetch(loggingFetch({ track, log }))
```

Since `loggingFetch` returns a fetch compatible function, `validatedLoggingFetch` should not care about the fact that it deals with a fetch that is augmented with logging capabilites. Instead make it accept a fetch instance as parameter. (Unix command design principle)

pudkrong replied:
I will use the existing function, [validatedFetch](https://github.com/hello-nrfcloud/backend/blob/291-implement-nrf-cloud-api-response-validation/nrfcloud/validatedFetch.ts), because

- It accepts `fetch` instance 
- However, I need to modify the function to accept 3rd parameter because the existing function will use `GET` method and `Content-Type` as `application/json` as its defaults.

coderbyheart commented on lambda/loggingFetch.ts:
```diff
@@ -21,3 +47,30 @@ export const loggingFetch =
 		return res
 	}
+
+export const validatedLoggingFetch = (
```
Move validation aspect to a separate file. (see below)

coderbyheart commented on lambda/resolveSingleCellGeoLocation.ts:
```diff
@@ -64,20 +62,17 @@ const serviceToken = once(
-			token: string // JWT
+
+		if ('error' in maybeResult) {
+			log.error('request failed', { error: maybeResult.error })
```
Make logging request errors part of validatingFetch, so it is not forgotte, and does not have to be repeated every time.

coderbyheart commented on lambda/resolveSingleCellGeoLocation.ts:
```diff
@@ -64,20 +62,17 @@ const serviceToken = once(
+
+		if ('error' in maybeResult) {
+			log.error('request failed', { error: maybeResult.error })
+			throw new Error(`Acquiring service token failed: ${maybeResult.error}`)
```
**comment:** this swallows the validation errors (because the error is stringified). But OK in this case because error is logged and function is only used in lambda.

coderbyheart commented on nrfcloud/createAccountDevice.ts:
```diff
@@ -12,18 +13,25 @@ export const createAccountDevice = async ({
+
+	if ('error' in maybeResult) {
+		console.error(`Failed to create account device:`, maybeResult.error)
+		throw new Error(`Failed to create account device`)
```
**problem:** this swallows the error details, change API and return error

pudkrong replied:
I will re-throw the error from `validatedFetch`. The error can be either `Error` object or `ValidationError`.

coderbyheart commented on nrfcloud/createAccountDevice.ts:
```diff
@@ -12,18 +13,25 @@ export const createAccountDevice = async ({
+	)
+
+	if ('error' in maybeResult) {
+		console.error(`Failed to create account device:`, maybeResult.error)
```
**problem:** Do not call console.log from functions.

coderbyheart commented on nrfcloud/devices.ts:
```diff
@@ -84,6 +84,10 @@ const Page = <T extends TSchema>(Item: T) =>
 	})
 const Devices = Page(Device)
 
+const ProvisionDevice = Type.Object({
+	bulkOpsRequestId: Type.String(),
+})
```
Add link to documentation.

coderbyheart commented on nrfcloud/getDeviceShadowFromnRFCloud.ts:
```diff
@@ -1,20 +1,38 @@
-import { type DeviceShadow } from './DeviceShadow.js'
+import { validatedFetch } from './validatedFetch.js'
+
+const DeviceShadow = Type.Object({
```
Add link to documentation

coderbyheart commented on nrfcloud/getDeviceShadowFromnRFCloud.ts:
```diff
@@ -1,20 +1,38 @@
+const DeviceShadow = Type.Object({
+	id: Type.String(),
+	state: Type.Object({
+		reported: Type.Record(Type.String(), Type.Any()),
+		version: Type.Number(),
+		metadata: Type.Record(Type.String(), Type.Any()),
```
Do not use any.

pudkrong replied:
Change to use `Type.Unknown()`

coderbyheart commented on nrfcloud/validatedFetch.ts:
```diff
@@ -31,7 +31,9 @@ const fetchData =
 		if (!response.ok)
-			throw new Error(`Error fetching status: ${response.status}`)
+			throw new Error(
+				`Error fetching status: ${response.status} - ${response.statusText}`,
```
Unrelated change.

coderbyheart commented on nrfcloud/validatedFetch.ts:
```diff
@@ -44,12 +46,14 @@ export const validatedFetch =
 	async <Schema extends TObject>(
 		{ resource }: { resource: string },
 		schema: Schema,
+		init?: RequestInit,
```
Unrelated change.

pudkrong replied:
This change is necessary due to the default behavior uses `GET` method and `Content-Type` as `application/json`

coderbyheart replied:
Changed in the refactor.

pudkrong committed:
feat: reuse validatedFetch

pudkrong commented on lambda/fetchDeviceShadow.ts:
```diff
@@ -215,7 +215,9 @@ const h = async (): Promise<void> => {
 					Math.round(Date.now() / 1000) -
-						getShadowUpdateTime(deviceShadow.state.metadata),
+						getShadowUpdateTime(
+							deviceShadow.state.metadata as Record<string, any>,
```
I need to cast `deviceShadow.state.metadata` to `Record<string, any>` because `getShadowUpdateTime` requires

coderbyheart replied:
Yeah, recursive typing is not really possible with TypeBox (and JSON schema)

pudkrong commented on lambda/loggingFetch.ts:
```diff
@@ -4,7 +4,10 @@ import type { Logger } from '@aws-lambda-powertools/logger'
 	({ track, log }: { track: AddMetricsFn; log: Logger }) =>
-	async (url: URL, init?: RequestInit): ReturnType<typeof fetch> => {
+	async (
+		url: URL | RequestInfo,
```
This is required to allow `loggingFetch` to be able to pass as `fetchImplementation`

coderbyheart replied:
I've addressed this in a refactor.

coderbyheart commented on lambda/validatedFetch.ts:
File is empty.

coderbyheart committed:
fix: remove empty file

coderbyheart commented on lambda/resolveSingleCellGeoLocation.ts:
```diff
@@ -153,9 +157,11 @@ const h = async (event: {
-		const res = await trackFetch(
-			new URL(`${slashless(apiEndpoint)}/v1/location/ground-fix`),
+
+		const vf = validatedFetch({ endpoint: apiEndpoint, apiKey }, trackFetch)
```
locationServiceToken can be passed here.

coderbyheart committed:
refactor: use higher-level API for validatedFetch

This updates the validateFetch API and integrates
the fetch configuration in the first parameter.

This way there are not two parameters (that
are separated by the schema parameter) that control
what fetch sends.

It also adds convenience functions to provide
payloads and sends the correct content-type header
only if needed.

coderbyheart commented on nrfcloud/getDeviceShadowFromnRFCloud.ts:
```diff
@@ -1,20 +1,41 @@
+/**
+ * @link https://api.nrfcloud.com/v1/#tag/All-Devices/operation/ListDevices
+ */
+const DeviceShadow = Type.Object({
```
Should have been integrated in DeviceShadow.ts

coderbyheart commented on nrfcloud/validatedFetch.spec.ts:
```diff
@@ -50,4 +50,41 @@ describe('validatedFetch()', () => {
+		)
+		const schema = Type.Object({ foo: Type.Literal('bar') })
+
+		const res = await vf({ resource: 'foo' }, schema, {
+			method: 'POST',
+			headers: {
+				Authorization: 'Bearer another-key',
+				'Content-Type': 'application/octet-stream',
+			},
+		})
```
If refactored this, because this mixes two parameters to configure the fetch behaviour. Those two (the first and the third) are also split around the schema.

coderbyheart committed:
refactor: improve DeviceShadow type and merge with existing type
